opm/simulators/wells/WellGroupHelpers.hpp

```diff
@@ -0,0 +1,211 @@
+
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            //only count producers or injectors
+            if ( (wellEcl.isProducer() && injector) ||  (wellEcl.isInjector() && !injector))
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void sumWellResRates(const Group2& group, const Schedule& schedule, const WellStateFullyImplicitBlackoil& wellState, const int reportStepIdx, const int phasePos, const bool injector, double& rate) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            sumWellResRates(groupTmp, schedule, wellState, reportStepIdx, phasePos, injector, rate);
+            rate *= groupTmp.getGroupEfficiencyFactor();
+        }
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
+                continue;
+
+            int well_index = it->second[0];
+
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            //only count producers or injectors
+            if ( (wellEcl.isProducer() && injector) ||  (wellEcl.isInjector() && !injector))
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void setGroupControl(const Group2& group, const Schedule& schedule, const int reportStepIdx, const bool injector, WellStateFullyImplicitBlackoil& wellState) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            setGroupControl(groupTmp, schedule, reportStepIdx, injector, wellState);
+            if (injector)
+                wellState.setCurrentInjectionGroupControl(groupName, "FLD");
+            else
+                wellState.setCurrentProductionGroupControl(groupName, "FLD");
+        }
+
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
```

Comment:
Here, and similar places in the code elsewhere, behaviour is unclear to me:
 - Either it is a bug, it should never happen.
 - Or, the state can be missing the well because it is shut or on another process.

I think the second applies, but that makes me worry how this works out in parallel? The existing group controls do not work well in parallel, but that is one of the things we must fix.

Reply:
> Or, the state can be missing the well because it is shut or on another process.

The well was missing for the UDQ test case. I haven't tested the UDQ testcase in parallell.

We need to communicate the wellstate for the parallell case. The well data structure from opm-common is already communicated in the output code. As I said earlier I hope we can replace the current wellState with the welldata structure from opm-common. Then we already have a code that communicates it. All process should know the full welldata from all processors.